urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/api/LauncherAPI.java

```diff
@@ -0,0 +1,335 @@
+                dir = new File(root, "Library/Application Support/" + appName);
+                break;
+            
+            case windows:
+                final String applicationData = System.getenv("APPDATA");
+                if (applicationData != null
+                        && !main.getConfig().getBoolean(
+                                "game.folder.customFolder"))
+                {
+                    root = new File(applicationData);
+                }
+                
+                dir = new File(root, '.' + appName + '/');
+                break;
+            
+            default:
+                dir = new File(root, appName + '/');
+                break;
+        }
+        
+        if (dir != null && !dir.exists() && !new File(dir, "bin").mkdirs())
+        {
+            throw new RuntimeException(
+                    "The working directory could not be created: " + dir);
+        }
+        
+        return dir;
+    }
+    
+    public String postUrl(String dest, Map<String, String> params)
+    {
+        return postUrl(dest, params, null);
+    }
+    
+    public String postUrl(String dest, Map<String, String> params,
+            String keyFile)
+    {
+        String result = null;
+        HttpURLConnection connection = null;
+        
+        try
+        {
+            final URL url = new URL(dest);
+            connection = (HttpURLConnection) url.openConnection();
+            
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type",
+                    "application/x-www-form-urlencoded");
+            
+            final StringBuffer sb = new StringBuffer();
+            final Iterator<Entry<String, String>> iterator = params.entrySet()
+                    .iterator();
+            while (iterator.hasNext())
+            {
+                final Entry<String, String> param = iterator.next();
+                
+                sb.append(param.getKey());
+                sb.append('=');
+                sb.append(param.getValue());
+                
+                if (iterator.hasNext())
+                {
+                    sb.append('&');
+                }
+            }
+            
+            connection.setRequestProperty("Content-Length",
+                    Integer.toString(sb.toString().getBytes().length));
+            connection.setRequestProperty("Content-Language", "en-US");
+            
+            connection.setUseCaches(false);
+            connection.setDoInput(true);
+            connection.setDoOutput(true);
+            
+            connection.connect();
+            
+            if (url.getProtocol().contains("https") && keyFile != null)
+            {
+                final Certificate[] certs = ((HttpsURLConnection) connection)
+                        .getServerCertificates();
+                
+                final byte[] bytes = new byte[294];
+                final DataInputStream dis = new DataInputStream(
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *dis*. The resource is referenced at lines 207, 210. The resource is closed at line 211. There are other execution paths that do not contain closure statements, for example, when DataInputStream.readFully throws an exception. Ensure *dis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).